plugin-server/src/main/ingestion-queues/session-recordings-consumer.ts

```diff
@@ -42,267 +43,218 @@ export const startSessionRecordingEventsConsumer = async ({
-                        const clickHouseRecord = createPerformanceEvent(
-                            messagePayload.uuid,
-                            team.id,
-                            messagePayload.distinct_id,
-                            event.properties || {}
-                        )
-
-                        pendingMessages.push(
-                            produce(
-                                producer,
-                                KAFKA_PERFORMANCE_EVENTS,
-                                Buffer.from(JSON.stringify(clickHouseRecord)),
-                                message.key
-                            )
-                        )
-                    } else {
-                        status.warn('⚠️', 'invalid_message', {
-                            reason: 'invalid_event_type',
-                            type: event.event,
-                            partition: batch.partition,
-                            offset: message.offset,
-                        })
-                        eventDroppedCounter
-                            .labels({
-                                event_type: 'session_recordings',
-                                drop_cause: 'invalid_event_type',
-                            })
-                            .inc()
-                    }
-                } catch (error) {
-                    status.error('⚠️', 'processing_error', {
-                        eventId: event.uuid,
-                        error: error,
+        if (team.session_recording_opt_in) {
+            try {
+                if (event.event === '$snapshot') {
+                    const clickHouseRecord = createSessionRecordingEvent(
+                        messagePayload.uuid,
+                        team.id,
+                        messagePayload.distinct_id,
+                        parseEventTimestamp(event as PluginEvent),
+                        event.ip,
+                        event.properties || {}
+                    )
+
+                    await produce(
+                        producer,
+                        KAFKA_CLICKHOUSE_SESSION_RECORDING_EVENTS,
+                        Buffer.from(JSON.stringify(clickHouseRecord)),
+                        message.key ? Buffer.from(message.key) : null
+                    )
+                } else if (event.event === '$performance_event') {
+                    const clickHouseRecord = createPerformanceEvent(
+                        messagePayload.uuid,
+                        team.id,
+                        messagePayload.distinct_id,
+                        event.properties || {}
+                    )
+
+                    await produce(
+                        producer,
+                        KAFKA_PERFORMANCE_EVENTS,
+                        Buffer.from(JSON.stringify(clickHouseRecord)),
+                        message.key ? Buffer.from(message.key) : null
+                    )
```

Comment:
- It's OK to `await` here because `consume` will call `eachMessage` in parallel for how many messages are allowed in flight right? My understanding is that this promise will need `.then(mark the offset as ok to commit)`, right? 

- If we fail, does the consumer handle retries to `eachMessage`?

- While I agree that this consumption style offers the best performance (by allowing more costly messages to not block the main flow), it opens us to the failure mode of one `eachMessage` failing repeatedly:

  - the consumer will eagerly process lots of messages after it, then hit the in-flight limit
  - the consumer then stops progressing, and we'll page on lag -> what observability do we offer to on-call?
  - restarting the consumer will create a big amount of duplicates (in-flight minus one). How will CH handle these?

This workload being pretty simple, the probability is pretty low (main error case would be output message too big and rejected by kafka?), so it's a right training ground. But it would get a lot trickier for a kafka->CH inserter service for example.